Fixed #27906 -- Fixed count of RootElement in test.html.count().

jacobtylerwalls opened the pull request:
ticket-27906

matthiask commented on tests/test_utils/tests.py:
```diff
@@ -773,6 +773,10 @@ def test_count(self):
         self.assertEqual(dom2.count(dom1), 1)
 
+        # children are a superset of the tested html
+        dom2 = parse_html('<div><p>foo</p><p>bar</p><p>foo</p><p>bar</p></div>')
```
Does it also work if you leave away the wrapping DIV? It isn't obvious to me that the issue from the ticket (`'<a/><b/>'` should be contained by `'<a/><b/><c/>'`) is being addressed by this change.

jacobtylerwalls replied:
I see the concern. Yes, with `<a/><b/><c/>` they are parsed as children of a `RootElement` which takes the place of the explicit `<div>` here, so it's the same handling. I can make that explicit by adding the test case from the ticket.

carltongibson replied:
I'm inclined to just add the test from the ticket here. 

```diff
diff --git a/tests/test_utils/tests.py b/tests/test_utils/tests.py
index 28cb64512f..7eb7302b79 100644
--- a/tests/test_utils/tests.py
+++ b/tests/test_utils/tests.py
@@ -782,6 +782,10 @@ class HTMLEqualTests(SimpleTestCase):
         dom2 = parse_html('<div><p>foo</p><p>bar</p><p>foo</p><p>bar</p></div>')
         self.assertEqual(dom2.count(dom1), 2)

+        # maybe a nice comment 
+        dom1 = parse_html('<a/><b/>')
+        dom2 = parse_html('<a/><b/><c/>')
+        self.assertEqual(dom2.count(dom1), 1)
+
     def test_parsing_errors(self):
         with self.assertRaises(AssertionError):
             self.assertHTMLEqual('<p>', '')```

matthiask commented on django/test/html.py:
```diff
@@ -96,6 +97,17 @@ def _count(self, element, count=True):
                     elif element in child:
                         return 1
             else:
+                if isinstance(element, RootElement) and element.children:
```
I think you should add an explanation what this does or at least add a reference to the ticket to help future readers understand what this branch achieves.

matthiask commented on tests/test_utils/tests.py:
```diff
@@ -734,6 +734,11 @@ def test_html_contain(self):
         dom1 = parse_html('<div><p>foo</p><p>bar</p></div>')
         self.assertIn(dom2, dom1)
 
+    def test_assertInHTML(self):
```
Minor nit: Please don't mix `snake_case` and `camelCase` in one identifier if possible. `test_html_contain_without_parents` maybe?

(This will probably come up when committers add their review.)

jacobtylerwalls replied:
I don't like doing that either but I thought I would have to since I'm unit testing the function `assertInHTML()`. Thoughts?

matthiask replied:
Yes, that may be a good reason. Let's wait and see.

carltongibson commented on tests/test_utils/tests.py:
```diff
@@ -773,6 +778,10 @@ def test_count(self):
         dom2 = parse_html('<div><p>foo</p><p>bar</p></div>')
         self.assertEqual(dom2.count(dom1), 1)
 
+        # children are a superset of the tested html
```
I'm finding this comment too opaque. Can I ask you to rephrase so it guides a fresh reader quickly/easily?

carltongibson commented on django/test/html.py:
```diff
@@ -96,6 +97,19 @@ def _count(self, element, count=True):
                     elif element in child:
                         return 1
             else:
+                # Look for a complete sequence of element's children.
+                # Refs #27906.
```
I think we could do with a better comment here (and/or in the block).

* Why do we need to check this case? 
* We're relying on `__eq__`
* But also there's the `# complete match`
* When we exit the `if` block, how have we progressed the search? 

It's not easy to see what's going on.

carltongibson commented on django/test/html.py:
```diff
@@ -96,6 +97,19 @@ def _count(self, element, count=True):
+                    if elem_child == child:
+                        elem_child_idx += 1
+                        # complete match
+                        if elem_child_idx == len(element.children):
+                            i += 1
+                            elem_child_idx = 0
```
We're missing tests here. I'd expect a failure with this diff: 

```diff
diff --git a/django/test/html.py b/django/test/html.py
index b2af0938ce..8c2543a5da 100644
--- a/django/test/html.py
+++ b/django/test/html.py
@@ -86,7 +86,6 @@ class Element:
             if self.children == element.children:
                 return 1
         i = 0
-        elem_child_idx = 0
         for child in self.children:
             # child is text content and element is also text content, then
             # make a simple "text" in "text"
@@ -100,16 +99,10 @@ class Element:
                 # Look for a complete sequence of element's children.
                 # Refs #27906.
                 if isinstance(element, RootElement) and element.children:
-                    elem_child = element.children[elem_child_idx]
+                    elem_child = element.children[0]
                     # start or continue match
                     if elem_child == child:
-                        elem_child_idx += 1
-                        # complete match
-                        if elem_child_idx == len(element.children):
-                            i += 1
-                            elem_child_idx = 0
-                    else:
-                        elem_child_idx = 0
+                        i += 1
                 i += child._count(element, count=count)
                 if not count and i:
                     return i
```

(c.f. _"Better comments..."_ above)

carltongibson reviewed (commented):
Hi @jacobtylerwalls. Thanks for this. Various comments. 

Beyond the missing coverage, most come down to my not finding the added code at all clear. Maybe that's just me but can we try to make it clearer for those (future selves included) who follow here?

jacobtylerwalls commented:
Thanks for the review!

- Improved comments throughout
- moved my new `if` branch below the existing code
- added a comment over the old code to clarify the difference
- moved all tests to `test_count()`
- added a test for no match found 
- added a test for a match found as both child and grandchild that would fail if someone turns my `if` branch to `elif`

carltongibson reviewed (approved):
Hi @jacobtylerwalls. Thanks for this. Looks good. I squashed, rebased, and pushed very minor edits. 👍
